koku/subs/tasks.py

```diff
@@ -0,0 +1,117 @@
+    return bool(
+        UNLEASH_CLIENT.is_enabled("cost-management.backend.subs-data-extraction", context, fallback_development_true)
+        or settings.ENABLE_SUBS_DEBUG
+        or check_subs_source_gate(schema_name)
+    )
+
+
+def enable_subs_messaging(schema_name: str) -> bool:
+    """Helper to determine if source is enabled for SUBS messaging."""
+    schema_name = convert_account(schema_name)
+    context = {"schema_name": schema_name}
+    LOG.info(log_json(msg="enable_subs_messaging context", context=context))
+    return bool(
+        UNLEASH_CLIENT.is_enabled("cost-management.backend.subs-data-messaging", context) or settings.ENABLE_SUBS_DEBUG
+    )
+
+
+@celery_app.task(name="subs.tasks.collect_subs_report_data_from_manifest", queue=SUBS_EXTRACTION_QUEUE)
+def collect_subs_report_data_from_manifest(reports_to_subs_summarize):
+    """Implement the functionality of the new task"""
+    reports = [report for report in reports_to_subs_summarize if report]
+    reports_deduplicated = [dict(t) for t in {tuple(d.items()) for d in reports}]
+    for report in reports_deduplicated:
+        schema_name = report.get("schema_name")
+        provider_type = report.get("provider_type")
+        provider_uuid = report.get("provider_uuid")
+        tracing_id = report.get("tracing_id", report.get("manifest_uuid", str(uuid.uuid4())))
+        context = {"schema": schema_name, "provider_type": provider_type, "provider_uuid": provider_uuid}
+        if provider_type not in SUBS_ACCEPTED_PROVIDERS:
+            LOG.info(log_json(tracing_id, msg="provider type not valid for subs processing", context=context))
+            continue
+        if not enable_subs_extraction(schema_name):
+            LOG.info(log_json(tracing_id, msg="subs processing not enabled for provider", context=context))
+            continue
+        if report.get("start") and report.get("end"):
+            LOG.debug(
+                log_json(
+                    tracing_id, msg="using start and end dates from the manifest for subs processing", context=context
+                )
+            )
+            start_date = parser.parse(report.get("start"))
+            end_date = parser.parse(report.get("end"))
+        else:
+            # GCP and OCI set report start and report end, AWS/Azure do not
+            date_tuple = get_start_and_end_from_manifest_id(report.get("manifest_id"))
+            if not date_tuple:
+                LOG.debug(log_json(tracing_id, msg="skipping report, no manifest found.", context=context))
+                continue
+            start_date, end_date = date_tuple
+        if start_date is None:
+            start_date = DateAccessor().today() - datetime.timedelta(days=2)
+        if end_date is None:
+            end_date = DateAccessor().today()
+        if isinstance(start_date, str):
+            start_date = datetime.datetime.strptime(start_date, "%Y-%m-%d")
+            end_date = datetime.datetime.strptime(end_date, "%Y-%m-%d")
```

Comment:
I still don't like any of this data logic here. Given a report, we should have a function that accepts that report and returns the start and end dates that we require. All this logic here is messy and hard to follow.

Reply:
I have reworked the date logic into a standalone function and reduced the complexity heavily, we now only get the start of the month associated with the manifest or start date and process based on month start since we either need to process all of the current month or the latest timestamp.

Reply:
Sweet, thanks.